numba_dpex/core/kernel_interface/func.py

```diff
@@ -6,100 +6,224 @@
+                name="DpexFunctionTemplateCache",
+                capacity=config.CACHE_SIZE,
+                pyfunc=self._pyfunc,
+            )
+        else:
+            self._cache = NullCache()
+        self._cache_hits = 0
+
+    @property
+    def cache(self):
+        """Cache accessor"""
+        return self._cache
+
+    @property
+    def cache_hits(self):
+        """Cache hit count accessor"""
+        return self._cache_hits
 
     def compile(self, args):
-        """Compile a dpex.func decorated Python function with the given
-        argument types.
+        """Compile a numba_dpex.func decorated Python function
+
+        Compile a dpex.func decorated Python function with the given
+        argument types. Each signature is compiled once by caching the
+        compiled function inside this object.
 
-        Each signature is compiled once by caching the compiled function inside
-        this object.
+        Args:
+            args (tuple): Function argument types in a tuple.
+
+        Returns:
+            numba.core.typing.templates.Signature: Signature of the
+                compiled result.
         """
-        if args not in self._compileinfos:
+
+        argtypes = [
+            dpex_target.typing_context.resolve_argument_type(arg)
+            for arg in args
+        ]
+        key = build_key(
+            tuple(argtypes),
+            self._pyfunc,
+            dpex_target.target_context.codegen(),
+        )
+        cres = self._cache.get(key)
+        if cres is None:
+            self._cache_hits += 1
             cres = compile_with_dpex(
-                pyfunc=self.py_func,
-                pyfunc_name=self.py_func.__name__,
+                pyfunc=self._pyfunc,
+                pyfunc_name=self._pyfunc.__name__,
                 return_type=None,
                 target_context=dpex_target.target_context,
                 typing_context=dpex_target.typing_context,
                 args=args,
                 is_kernel=False,
-                debug=self.debug,
+                debug=self._debug,
             )
             func = cres.library.get_function(cres.fndesc.llvm_func_name)
             cres.target_context.mark_ocl_device(func)
-            first_definition = not self._compileinfos
-            self._compileinfos[args] = cres
             libs = [cres.library]
 
-            if first_definition:
-                # First definition
-                cres.target_context.insert_user_function(
-                    self, cres.fndesc, libs
-                )
-            else:
-                cres.target_context.add_user_function(self, cres.fndesc, libs)
+            cres.target_context.insert_user_function(self, cres.fndesc, libs)
+            self._cache.put(key, cres)
+        return cres.signature
 
-        else:
-            cres = self._compileinfos[args]
 
-        return cres.signature
+def compile_func(pyfunc, signature, debug=None):
+    """Compile a DpexFunction
```

Comment:
Docstring needs fixing, How about:

"Compiles a specialized `numba_dpex.func` decorated function to native binary library function and returns the library wrapped inside a :class:`numba_dpex.core.kernel_interface.func.DpexFunction` object."